### PR TITLE
Scheduler: fix calculateBackoffDuration overflow in extreme data cases

### PR DESCRIPTION
**Pull request by binacs**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:
Fixes #https://github.com/kubernetes/kubernetes/issues/105094

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

Thanks for review ^-^

### CODE REVIEW

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 1]
```diff
@@ -742,10 +742,10 @@ func (p *PriorityQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Ti
 func (p *PriorityQueue) calculateBackoffDuration(podInfo *framework.QueuedPodInfo) time.Duration {
 	duration := p.podInitialBackoffDuration
 	for i := 1; i < podInfo.Attempts; i++ {
-		duration = duration * 2
-		if duration > p.podMaxBackoffDuration {
+		if duration > p.podMaxBackoffDuration-duration {
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue.go:745]: Add a comment explaining the intent.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 2]
```diff
@@ -1995,3 +1995,56 @@ func makeQueuedPodInfos(num int, timestamp time.Time) []*framework.QueuedPodInfo
...
 }
+
+func TestPriorityQueue_calculateBackoffDuration(t *testing.T) {
+	type args struct {
+		podInfo *framework.QueuedPodInfo
+	}
+	tests := []struct {
+		name  string
+		queue *PriorityQueue
+		args  args
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2006]: Directly use podInfo as the field:

```suggestion
		podInfo *framework.QueuedPodInfo
```

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 3]
```diff
@@ -1995,3 +1995,56 @@ func makeQueuedPodInfos(num int, timestamp time.Time) []*framework.QueuedPodInfo
...
 	return pInfos
 }
+
+func TestPriorityQueue_calculateBackoffDuration(t *testing.T) {
+	type args struct {
+		podInfo *framework.QueuedPodInfo
+	}
+	tests := []struct {
+		name  string
+		queue *PriorityQueue
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2005]: Replace it with `initialBackoffDuration` and `maxBackoffDuration`, and then instantiate the queue object in the testing block.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 4]
```diff
@@ -1995,3 +1995,56 @@ func makeQueuedPodInfos(num int, timestamp time.Time) []*framework.QueuedPodInfo
...
+					PodInfo:  nil,
+					Attempts: 64,
+				},
+			},
+			want: math.MaxInt64 * time.Nanosecond,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := tt.queue.calculateBackoffDuration(tt.args.podInfo); !reflect.DeepEqual(got, tt.want) {
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2045]:
```suggestion
			if got := tt.queue.calculateBackoffDuration(tt.args.podInfo); got != tt.want {
```

**Comment by BinacsLee** [comment]
Done, PTAL! @Huang-Wei  ^-^

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 5]
```diff
@@ -742,10 +742,12 @@ func (p *PriorityQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Ti
 func (p *PriorityQueue) calculateBackoffDuration(podInfo *framework.QueuedPodInfo) time.Duration {
 	duration := p.podInitialBackoffDuration
 	for i := 1; i < podInfo.Attempts; i++ {
-		duration = duration * 2
-		if duration > p.podMaxBackoffDuration {
+		// We need to check whether the duration's twice is greater than the podMaxBackoffDuration.
+		// And we use subtraction instead of addition or multiplication to avoid numerical overflow.
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue.go:746]: Simplify the wordings:

```
// Use subtraction instead of addition or multiplication to avoid overflow.
```

**Review by Huang-Wei — COMMENTED** [review]
LGTM overall. Please squash the commits along with resolving the last comment.

**Comment by BinacsLee** [comment]
Done, thanks for review.
